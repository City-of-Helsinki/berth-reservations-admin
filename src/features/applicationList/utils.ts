import {
  BERTH_APPLICATIONS,
  BERTH_APPLICATIONS_berthApplications_edges as BERTH_APPLICATIONS_EDGE,
  BERTH_APPLICATIONS_berthApplications_edges_node as BERTH_APPLICATIONS_NODE,
} from './__generated__/BERTH_APPLICATIONS';
import { ApplicationStatus, LeaseStatus } from '../../@types/__generated__/globalTypes';

interface HarborChoice {
  harbor: string;
  harborName: string;
  priority: number;
}

interface Lease {
  berthNum: string;
  harborId: string;
  harborName: string;
  id: string;
  pierIdentifier: string;
  status: LeaseStatus;
}

interface BerthSwitch {
  berthNum: string;
  harborId: string;
  harborName: string;
  pierIdentifier: string;
  reason: string | null;
}

export interface ApplicationData {
  accessibilityRequired: boolean;
  berthSwitch: BerthSwitch | null;
  boatDraught: number | null;
  boatLength: number;
  boatModel: string;
  boatName: string;
  boatRegistrationNumber: string;
  boatType?: string | null;
  boatWeight: number | null;
  boatWidth: number;
  choices: Array<HarborChoice>;
  createdAt: string;
  customerId?: string;
  firstName: string;
  id: string;
  isSwitch: boolean;
  lastName: string;
  lease: Lease | null;
  municipality: string;
  queue: number;
  status: ApplicationStatus;
}

export const getBerthApplicationData = (data: BERTH_APPLICATIONS | undefined): ApplicationData[] => {
  const boatTypes = data?.boatTypes;

  return (
    data?.berthApplications?.edges.reduce<ApplicationData[]>((acc, edge) => {
      const {
        accessibilityRequired,
        berthSwitch,
        boatDraught,
        boatLength,
        boatModel,
        boatName,
        boatRegistrationNumber,
        boatType,
        boatWeight,
        boatWidth,
        createdAt,
        customer,
        firstName,
        harborChoices,
        id,
        lastName,
        lease,
        municipality,
        status,
      } = (edge as BERTH_APPLICATIONS_EDGE).node as BERTH_APPLICATIONS_NODE;

      const choices =
        harborChoices?.map((choice) => {
          return {
            priority: choice?.priority ?? Number.MAX_VALUE,
            harbor: choice?.harbor ?? '',
            harborName: choice?.harborName ?? '',
          };
        }) ?? [];

      let leaseProps: Lease | null = null;
      if (lease?.berth?.pier.properties?.harbor) {
        leaseProps = {
          berthNum: lease.berth.number || '',
          harborId: lease.berth.pier.properties.harbor.id,
          harborName: lease.berth.pier.properties.harbor.properties?.name || '',
          id: lease.id,
          pierIdentifier: lease.berth.pier.properties?.identifier || '',
          status: lease.status,
        };
      }

      const berthSwitchProps = berthSwitch && {
        berthNum: berthSwitch.berthNumber,
        harborId: berthSwitch.harbor,
        harborName: berthSwitch.harborName,
        pierIdentifier: berthSwitch.pier,
        reason: berthSwitch.reason?.title || null,
      };

      const applicationData = {
        accessibilityRequired,
        berthSwitch: berthSwitchProps,
        boatDraught,
        boatLength,
        boatModel,
        boatName,
        boatRegistrationNumber,
        boatType: boatTypes?.find(({ id }) => id === boatType)?.name,
        boatWeight,
        boatWidth,
        choices,
        createdAt,
        customerId: customer?.id,
        firstName,
        id,
        isSwitch: !!berthSwitch,
        lastName,
        lease: leaseProps,
        municipality,
        queue: 0, // TODO
        status,
      };

      return [...acc, applicationData];
    }, []) ?? []
  );
};
