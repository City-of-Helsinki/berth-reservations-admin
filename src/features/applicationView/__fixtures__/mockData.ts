import {
  INDIVIDUAL_APPLICATION,
  INDIVIDUAL_APPLICATION_additionalProducts as ADDITIONAL_PRODUCTS,
  INDIVIDUAL_APPLICATION_berthApplication as BERTH_APPLICATION,
  INDIVIDUAL_APPLICATION_berthApplication_berthSwitch as BERTH_SWITCH,
  INDIVIDUAL_APPLICATION_berthApplication_customer as CUSTOMER,
  INDIVIDUAL_APPLICATION_berthApplication_lease as LEASE,
  INDIVIDUAL_APPLICATION_boatTypes as BOAT_TYPES,
} from '../__generated__/INDIVIDUAL_APPLICATION';
import {
  AdditionalProductType,
  ApplicationStatus,
  BerthApplicationLanguage,
  BerthMooringType,
  CustomerGroup,
  InvoicingType,
  Language,
  PeriodType,
  PriceUnits,
  ProductServiceType,
} from '../../../@types/__generated__/globalTypes';

export const mockCustomer: CUSTOMER = {
  __typename: 'ProfileNode',
  comment: 'Comment',
  firstName: 'Testi',
  id: 'MOCK-CUSTOMER',
  invoicingType: InvoicingType.DIGITAL_INVOICE,
  language: Language.FINNISH,
  lastName: 'Testinen',
  organization: null,
  primaryAddress: { __typename: 'AddressNode', address: 'Testikatu 1', city: 'Helsinki', postalCode: '00100' },
  primaryEmail: { __typename: 'EmailNode', email: 'test@example.com' },
  primaryPhone: { __typename: 'PhoneNode', phone: '+358 00 000 0000' },
  customerGroup: CustomerGroup.PRIVATE,
};

export const mockBerthSwitch: BERTH_SWITCH = {
  __typename: 'BerthSwitchType',
  berthNumber: '1',
  harbor: 'MOCK-HARBOR',
  harborName: 'Test Harbor',
  id: 'MOCK-BERTH-SWITCH',
  pier: 'A',
  reason: { __typename: 'BerthSwitchReasonType', title: 'Reason' },
};

export const mockLease: LEASE = {
  __typename: 'BerthLeaseNode',
  berth: {
    __typename: 'BerthNode',
    comment: '',
    depth: null,
    isAccessible: null,
    length: 5,
    mooringType: BerthMooringType.DINGHY_PLACE,
    number: 1,
    pier: {
      __typename: 'PierNode',
      properties: {
        __typename: 'PierProperties',
        electricity: false,
        gate: false,
        harbor: {
          __typename: 'HarborNode',
          id: 'MOCK-HARBOR',
          properties: { __typename: 'HarborProperties', name: 'Test Harbor' },
        },
        identifier: 'A',
        lighting: false,
        mooring: false,
        wasteCollection: false,
        water: false,
      },
    },
    width: 0,
  },
  id: 'MOCK-LEASE',
  order: null,
};

const additionalProducts: ADDITIONAL_PRODUCTS = {
  __typename: 'AdditionalProductNodeConnection',
  edges: [
    {
      __typename: 'AdditionalProductNodeEdge',
      node: {
        __typename: 'AdditionalProductNode',
        id: 'MOCK-ADDITIONAL-PRODUCT-NODE',
        period: PeriodType.MONTH,
        priceUnit: PriceUnits.AMOUNT,
        priceValue: '25.00',
        productType: AdditionalProductType.OPTIONAL_SERVICE,
        service: ProductServiceType.DINGHY_PLACE,
      },
    },
  ],
};

const berthApplication: BERTH_APPLICATION = {
  __typename: 'BerthApplicationNode',
  accessibilityRequired: false,
  address: 'Testikatu 1',
  berthSwitch: null,
  boatDraught: 1,
  boatLength: 1,
  boatModel: 'Marine',
  boatName: 'Cama la Yano',
  boatRegistrationNumber: '1234',
  boatType: '7',
  boatWeight: 120,
  boatWidth: 1,
  businessId: '',
  companyName: '',
  createdAt: '2020-07-10T08:04:36.062819+00:00',
  customer: null,
  email: 'test@example.com',
  firstName: 'Testi',
  harborChoices: [
    { harbor: 'MOCK-HARBOR-B', priority: 1, harborName: 'Saukonpaaden venesatama', __typename: 'HarborChoiceType' },
  ],
  id: 'MOCK-APPLICATION',
  language: BerthApplicationLanguage.FI,
  lastName: 'Testinen',
  lease: null,
  municipality: 'Helsinki',
  phoneNumber: '+358 00 000 0000',
  status: ApplicationStatus.PENDING,
  zipCode: '00100',
  acceptBoatingNewsletter: true,
  acceptFitnessNews: false,
  acceptLibraryNews: false,
  acceptOtherCultureNews: true,
  applicationCode: '1234',
};

const boatTypes: BOAT_TYPES[] = [
  { id: '1', name: 'Jollavene', __typename: 'BoatTypeType' },
  { id: '2', name: 'Soutuvene', __typename: 'BoatTypeType' },
  { id: '3', name: 'Perämoottorivene', __typename: 'BoatTypeType' },
  { id: '4', name: 'Sisäperämoottorivene', __typename: 'BoatTypeType' },
  { id: '5', name: 'Keskimoottorivene', __typename: 'BoatTypeType' },
  { id: '6', name: 'Purjevene / moottoripursi', __typename: 'BoatTypeType' },
  { id: '7', name: 'Troolari', __typename: 'BoatTypeType' },
  { id: '8', name: 'Suuri alus (yli 20t)', __typename: 'BoatTypeType' },
];

export const mockApplication: INDIVIDUAL_APPLICATION = {
  __typename: 'Query',
  additionalProducts: additionalProducts,
  berthApplication: berthApplication,
  boatTypes: boatTypes,
};
