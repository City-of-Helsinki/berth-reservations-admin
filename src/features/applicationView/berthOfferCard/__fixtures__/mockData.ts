import { BerthLease_lease as BerthLease } from '../__generated__/BerthLease';
import {
  AdditionalProductType,
  BerthMooringType,
  LeaseStatus,
  ProductServiceType,
} from '../../../../@types/__generated__/globalTypes';

export const lease: BerthLease = {
  id: '165023f5-5b74-4656-9315-d4e1a22f111d',
  status: LeaseStatus.DRAFTED,
  berth: {
    depth: null,
    length: 6,
    mooringType: BerthMooringType.SINGLE_SLIP_PLACE,
    width: 2.5,
    comment: '',
    isAccessible: null,
    number: '1',
    pier: {
      properties: {
        identifier: 'C',
        electricity: true,
        gate: true,
        lighting: true,
        mooring: true,
        wasteCollection: true,
        water: true,
        harbor: {
          id: 'b55da996-cf00-4b52-9aa0-6b13fd8d806b',
          properties: {
            name: 'Aurinkolahden venesatama (Aurinkoranta)',
            __typename: 'HarborProperties',
          },
          __typename: 'HarborNode',
        },
        __typename: 'PierProperties',
      },
      __typename: 'PierNode',
    },
    __typename: 'BerthNode',
  },
  customer: {
    __typename: 'ProfileNode',
    id: 'MOCK-CUSTOMER',
    primaryEmail: {
      __typename: 'EmailNode',
      id: 'MOCK-EMAIL',
      email: 'test@example.com',
    },
  },
  order: {
    id: '07f65f54-4a25-47fd-9eea-d9fdddfa46f9',
    orderNumber: '123456',
    price: '96.67',
    totalPrice: '256.70',
    orderLines: {
      edges: [
        {
          node: {
            id: '9b97b9bc-bba5-46d9-aae2-86e8686012ef',
            price: '20.50',
            product: {
              id: 'dc7f8158-a4cb-49a9-9368-e586f224fdb5',
              service: ProductServiceType.ELECTRICITY,
              productType: AdditionalProductType.FIXED_SERVICE,
              __typename: 'AdditionalProductNode',
            },
            __typename: 'OrderLineNode',
          },
          __typename: 'OrderLineNodeEdge',
        },
        {
          node: {
            id: 'ea2ef175-f790-45a0-b16f-9e5a98340691',
            price: '19.33',
            product: {
              id: '2e55f605-061e-4299-814d-bf2cd53aa8ce',
              service: ProductServiceType.WATER,
              productType: AdditionalProductType.FIXED_SERVICE,
              __typename: 'AdditionalProductNode',
            },
            __typename: 'OrderLineNode',
          },
          __typename: 'OrderLineNodeEdge',
        },
        {
          node: {
            id: '8ba4de61-63cc-49fc-afe2-26068f3c8e4d',
            price: '2.42',
            product: {
              id: '3a2e3a2f-2782-40e9-a139-b9a4d4a36c3b',
              service: ProductServiceType.GATE,
              productType: AdditionalProductType.FIXED_SERVICE,
              __typename: 'AdditionalProductNode',
            },
            __typename: 'OrderLineNode',
          },
          __typename: 'OrderLineNodeEdge',
        },
        {
          node: {
            id: '59c768c1-f371-4178-afda-1e16d2c725b0',
            price: '8.33',
            product: {
              id: '762da6e9-89be-43bc-b097-53dff4ed8415',
              service: ProductServiceType.MOORING,
              productType: AdditionalProductType.FIXED_SERVICE,
              __typename: 'AdditionalProductNode',
            },
            __typename: 'OrderLineNode',
          },
          __typename: 'OrderLineNodeEdge',
        },
        {
          node: {
            id: '6d910d8e-80b6-4486-aa85-4958bcd98dd5',
            price: '4.17',
            product: {
              id: '34c5b6bc-7eb8-457e-bce8-8ba3a07b9057',
              service: ProductServiceType.WASTE_COLLECTION,
              productType: AdditionalProductType.FIXED_SERVICE,
              __typename: 'AdditionalProductNode',
            },
            __typename: 'OrderLineNode',
          },
          __typename: 'OrderLineNodeEdge',
        },
        {
          node: {
            id: 'ab7cf1e5-92b5-42ad-844a-e2d577d30baf',
            price: '40.28',
            product: {
              id: '650dbfcb-8f10-46b6-af64-197b77686a5b',
              service: ProductServiceType.LIGHTING,
              productType: AdditionalProductType.FIXED_SERVICE,
              __typename: 'AdditionalProductNode',
            },
            __typename: 'OrderLineNode',
          },
          __typename: 'OrderLineNodeEdge',
        },
        {
          node: {
            id: '244a7af9-6f91-4f08-9738-82ef06a2e790',
            price: '65.00',
            product: {
              id: '5194dd14-0226-4603-808c-831cfbec72de',
              service: ProductServiceType.DINGHY_PLACE,
              productType: AdditionalProductType.OPTIONAL_SERVICE,
              __typename: 'AdditionalProductNode',
            },
            __typename: 'OrderLineNode',
          },
          __typename: 'OrderLineNodeEdge',
        },
      ],
      __typename: 'OrderLineNodeConnection',
    },
    __typename: 'OrderNode',
  },
  __typename: 'BerthLeaseNode',
};
