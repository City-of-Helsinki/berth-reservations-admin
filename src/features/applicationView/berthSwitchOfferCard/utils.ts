import { INDIVIDUAL_APPLICATION_berthApplication_switchOffers as BerthSwitchOffersData } from '../__generated__/INDIVIDUAL_APPLICATION';
import { BerthSwitchOfferDetails } from './types';

export const getSwitchOffers = (switchOffersData: BerthSwitchOffersData): BerthSwitchOfferDetails[] => {
  return switchOffersData.edges.reduce<BerthSwitchOfferDetails[]>((acc, edge) => {
    const switchOffer = edge?.node;
    if (switchOffer) {
      const berthSwitchOffer = {
        id: switchOffer.id,
        status: switchOffer.status,
        berthComment: switchOffer.berth.comment,
        berthDepth: switchOffer.berth.depth,
        berthIsAccessible: switchOffer.berth.isAccessible ?? false,
        berthLength: switchOffer.berth.length,
        berthMooringType: switchOffer.berth.mooringType,
        berthNum: switchOffer.berth.number,
        berthWidth: switchOffer.berth.width,
        electricity: switchOffer.berth.pier.properties?.electricity ?? false,
        gate: switchOffer.berth.pier.properties?.gate ?? false,
        harborName: switchOffer.berth.pier.properties?.harbor.properties?.name ?? '',
        lighting: switchOffer.berth.pier.properties?.lighting ?? false,
        pierIdentifier: switchOffer.berth.pier.properties?.identifier ?? '',
        wasteCollection: switchOffer.berth.pier.properties?.wasteCollection ?? false,
        water: switchOffer.berth.pier.properties?.water ?? false,
      };
      return [...acc, berthSwitchOffer];
    }
    return acc;
  }, []);
};
