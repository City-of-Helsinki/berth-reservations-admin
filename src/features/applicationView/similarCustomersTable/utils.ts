import { FILTERED_CUSTOMERS } from './__generated__/FILTERED_CUSTOMERS';
import { CustomerData } from '../ApplicationView';

export const getFilteredCustomersData = (data?: FILTERED_CUSTOMERS): CustomerData[] => {
  if (!data?.profiles) return [];

  return data.profiles.edges.reduce<CustomerData[]>((acc, edge) => {
    if (!edge?.node) return acc;
    const { id, firstName, lastName, primaryAddress, berthLeases, customerGroup } = edge.node;

    const berths = berthLeases?.edges
      .map((edge) => edge?.node?.berth?.pier.properties?.harbor.properties?.name)
      .join(', ');

    return [
      ...acc,
      {
        id,
        name: `${lastName}, ${firstName}`,
        city: primaryAddress?.city,
        address: primaryAddress?.address,
        berths,
        customerGroup,
      },
    ];
  }, []);
};
