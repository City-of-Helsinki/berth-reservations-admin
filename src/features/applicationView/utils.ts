import { ApplicationDetailsProps } from '../../common/applicationDetails/ApplicationDetails';
import {
  INDIVIDUAL_APPLICATION_berthApplication as BERTH_APPLICATION,
  INDIVIDUAL_APPLICATION_berthApplication_customer as CUSTOMER_PROFILE,
  INDIVIDUAL_APPLICATION_boatTypes as BOAT_TYPES,
} from './__generated__/INDIVIDUAL_APPLICATION';
import { CustomerProfileCardProps } from '../../common/customerProfileCard/CustomerProfileCard';
import { CustomerData } from './ApplicationView';
import { BerthApplicationLanguage, CustomerGroup, Language } from '../../@types/__generated__/globalTypes';
import { FILTERED_CUSTOMERS } from './__generated__/FILTERED_CUSTOMERS';

export const getCustomerProfile = (profile: CUSTOMER_PROFILE): CustomerProfileCardProps => {
  return {
    customerId: profile.id,
    firstName: profile.firstName,
    lastName: profile.lastName,
    primaryAddress: profile.primaryAddress,
    primaryPhone: profile.primaryPhone?.phone,
    primaryEmail: profile.primaryEmail?.email,
    ssn: '-', // TODO
    language: profile.language,
    showCustomerNameAsLink: true,
    customerGroup: profile.customerGroup,
  };
};

interface Lease {
  berthNum: string | number;
  harborId: string;
  harborName: string;
  id: string;
  pierIdentifier: string;
}

interface BerthSwitch {
  berthNum: string;
  harborId: string;
  harborName: string;
  pierIdentifier: string;
  reason: string | null;
}

const mapBerthApplicationLanguageToLanguage = (berthApplicationLanguage: BerthApplicationLanguage): Language | null => {
  switch (berthApplicationLanguage) {
    case BerthApplicationLanguage.FI:
      return Language.FINNISH;
    case BerthApplicationLanguage.SV:
      return Language.SWEDISH;
    case BerthApplicationLanguage.EN:
      return Language.ENGLISH;
    default:
      return null;
  }
};

const getApplicantDetails = (berthApplication: BERTH_APPLICATION): CustomerProfileCardProps => {
  const {
    firstName,
    lastName,
    address,
    zipCode,
    municipality,
    phoneNumber,
    email,
    businessId,
    companyName,
    language,
  } = berthApplication;
  const customerGroup = businessId ? CustomerGroup.COMPANY : CustomerGroup.PRIVATE;

  return {
    firstName: firstName,
    lastName: lastName,
    primaryAddress: {
      address: address,
      postalCode: zipCode,
      city: municipality,
    },
    primaryPhone: phoneNumber,
    primaryEmail: email,
    language: mapBerthApplicationLanguageToLanguage(language),
    customerGroup,
    ...(businessId && {
      organization: {
        businessId,
        name: companyName,
        address,
        city: municipality,
        postalCode: zipCode,
      },
    }),
  };
};

export const getApplicationDetailsData = (
  berthApplication: BERTH_APPLICATION,
  boatTypes: BOAT_TYPES[]
): ApplicationDetailsProps & Required<Pick<ApplicationDetailsProps, 'applicant'>> => {
  const harborChoices = berthApplication.harborChoices || [];
  const lease: Lease | null = berthApplication.lease
    ? {
        harborId: berthApplication.lease.berth?.pier.properties?.harbor.id || '',
        harborName: berthApplication.lease.berth?.pier.properties?.harbor.properties?.name || '',
        id: berthApplication.lease.id,
        berthNum: berthApplication.lease.berth?.number.toString(10) || '',
        pierIdentifier: berthApplication.lease.berth?.pier.properties?.identifier || '',
      }
    : null;
  const berthSwitch: BerthSwitch | null = berthApplication.berthSwitch
    ? {
        harborId: berthApplication.berthSwitch.harbor,
        harborName: berthApplication.berthSwitch.harborName,
        berthNum: berthApplication.berthSwitch.berthNumber,
        pierIdentifier: berthApplication.berthSwitch.pier,
        reason: berthApplication.berthSwitch.reason?.title || null,
      }
    : null;

  return {
    ...berthApplication,
    customerId: berthApplication.customer?.id,
    applicant: getApplicantDetails(berthApplication),
    berthSwitch,
    queue: null,
    harborChoices,
    lease,
    boatType: boatTypes.find(({ id }) => id === berthApplication.boatType)?.name,
  };
};

export const getFilteredCustomersData = (data?: FILTERED_CUSTOMERS): CustomerData[] => {
  if (!data?.profiles) return [];

  return data.profiles.edges.reduce<CustomerData[]>((acc, edge) => {
    if (!edge?.node) return acc;
    const { id, firstName, lastName, primaryAddress, berthLeases, customerGroup } = edge.node;

    const berths = berthLeases?.edges
      .map((edge) => edge?.node?.berth?.pier.properties?.harbor.properties?.name)
      .join(', ');

    return [
      ...acc,
      {
        id,
        name: `${lastName}, ${firstName}`,
        city: primaryAddress?.city,
        address: primaryAddress?.address,
        berths,
        customerGroup,
      },
    ];
  }, []);
};
