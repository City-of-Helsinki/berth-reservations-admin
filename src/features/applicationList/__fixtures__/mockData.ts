import { BERTH_APPLICATIONS } from '../__generated__/BERTH_APPLICATIONS';
import { getBerthApplicationData } from '../utils';
import { ApplicationStatus, LeaseStatus } from '../../../@types/__generated__/globalTypes';

export const mockData: BERTH_APPLICATIONS = {
  berthApplications: {
    __typename: 'BerthApplicationNodeConnection',
    count: 3,
    edges: [
      // Full structure
      {
        __typename: 'BerthApplicationNodeEdge',
        node: {
          __typename: 'BerthApplicationNode',
          accessibilityRequired: false,
          berthSwitch: {
            __typename: 'BerthSwitchType',
            berthNumber: '1',
            harbor: 'MOCK-HARBOR-0',
            harborName: 'Testisatama',
            id: 'MOCK-BERTH-SWITCH-0',
            pier: 'A',
            reason: { __typename: 'BerthSwitchReasonType', title: 'Reason' },
          },
          boatLength: 6,
          boatModel: 'Marine',
          boatName: 'Cama la Yano',
          boatDraught: 5,
          boatRegistrationNumber: 'A 12345',
          boatType: '1',
          boatWeight: 5,
          boatWidth: 0,
          createdAt: '2020-07-16',
          customer: { __typename: 'ProfileNode', id: 'MOCK-PROFILE' },
          firstName: 'Matti',
          harborChoices: [{ __typename: 'HarborChoiceType', harbor: '1', harborName: '1', priority: 0 }],
          id: 'MOCK-APPLICATION-0',
          lastName: 'Meikäläinen',
          lease: {
            __typename: 'BerthLeaseNode',
            berth: {
              __typename: 'BerthNode',
              number: '7',
              pier: {
                __typename: 'PierNode',
                properties: {
                  __typename: 'PierProperties',
                  harbor: {
                    __typename: 'HarborNode',
                    id: 'MOCK-HARBOR-0',
                    properties: { __typename: 'HarborProperties', name: 'Testisatama' },
                  },
                  identifier: 'G',
                },
              },
            },
            id: 'MOCK-LEASE-0',
            status: LeaseStatus.OFFERED,
          },
          municipality: 'Helsinki',
          status: ApplicationStatus.PENDING,
        },
      },
      // Most structure and nulled fields
      {
        __typename: 'BerthApplicationNodeEdge',
        node: {
          __typename: 'BerthApplicationNode',
          accessibilityRequired: false,
          berthSwitch: {
            __typename: 'BerthSwitchType',
            berthNumber: '2',
            harbor: 'MOCK-HARBOR-0',
            harborName: 'Testisatama',
            id: 'MOCK-BERTH-SWITCH-1',
            pier: 'A',
            reason: null,
          },
          boatLength: 6,
          boatModel: 'Marine',
          boatName: 'Cama la Yano II',
          boatDraught: 5,
          boatRegistrationNumber: 'B 12345',
          boatType: '1',
          boatWeight: 5,
          boatWidth: 0,
          createdAt: '2020-07-16',
          customer: { __typename: 'ProfileNode', id: 'MOCK-PROFILE' },
          firstName: 'Maija',
          harborChoices: null,
          id: 'MOCK-APPLICATION-1',
          lastName: 'Meikäläinen',
          lease: {
            __typename: 'BerthLeaseNode',
            id: 'MOCK-LEASE-1',
            status: LeaseStatus.OFFERED,
            berth: {
              __typename: 'BerthNode',
              number: '7',
              pier: {
                __typename: 'PierNode',
                properties: {
                  __typename: 'PierProperties',
                  identifier: 'Pier Brosnan',
                  harbor: {
                    __typename: 'HarborNode',
                    id: 'MOCK-HARBOR-1',
                    properties: {
                      __typename: 'HarborProperties',
                      name: null,
                    },
                  },
                },
              },
            },
          },
          municipality: 'Helsinki',
          status: ApplicationStatus.PENDING,
        },
      },
      // Minimum fields
      {
        __typename: 'BerthApplicationNodeEdge',
        node: {
          __typename: 'BerthApplicationNode',
          accessibilityRequired: false,
          berthSwitch: null,
          boatLength: 6,
          boatModel: 'Marine',
          boatName: 'Cama la Yano 3-D',
          boatDraught: 5,
          boatRegistrationNumber: 'C 12345',
          boatType: '1',
          boatWeight: 5,
          boatWidth: 0,
          createdAt: '2020-07-16',
          customer: { __typename: 'ProfileNode', id: 'MOCK-PROFILE-0' },
          firstName: 'Matti',
          harborChoices: [
            { __typename: 'HarborChoiceType', harbor: 'MOCK-HARBOR-0', harborName: 'Testisatama', priority: 0 },
          ],
          id: 'MOCK-APPLICATION-2',
          lastName: 'Möttönen',
          lease: null,
          municipality: 'Helsinki',
          status: ApplicationStatus.PENDING,
        },
      },
    ],
  },
  boatTypes: [{ __typename: 'BoatTypeType', id: '1', name: 'Purjevene' }],
};

export const mockTableData = getBerthApplicationData(mockData);
