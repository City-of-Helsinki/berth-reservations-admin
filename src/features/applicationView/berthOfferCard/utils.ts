import { BerthLease } from './__generated__/BerthLease';
import { LeaseDetails } from './types';
import { getOrder } from '../../invoiceCard/utils';

export const getOfferDetailsData = (lease: BerthLease | null | undefined): LeaseDetails | null => {
  if (!lease) return null;

  const order: LeaseDetails['order'] = lease.order ? getOrder(lease.order) : null;

  return {
    id: lease.id,
    status: lease.status,
    berthComment: lease.berth?.comment ?? '',
    berthDepth: lease.berth?.depth ?? null,
    berthIsAccessible: lease.berth?.isAccessible ?? false,
    berthLength: lease.berth?.length ?? null,
    berthMooringType: lease.berth?.mooringType ?? null,
    berthNum: lease.berth?.number ?? '',
    berthWidth: lease.berth?.width ?? null,
    customerEmail: lease.customer.primaryEmail?.email ?? null,
    electricity: lease.berth?.pier.properties?.electricity ?? false,
    gate: lease.berth?.pier.properties?.gate ?? false,
    harborName: lease.berth?.pier.properties?.harbor.properties?.name ?? '',
    lighting: lease.berth?.pier.properties?.lighting ?? false,
    pierIdentifier: lease.berth?.pier.properties?.identifier ?? '',
    wasteCollection: lease.berth?.pier.properties?.wasteCollection ?? false,
    water: lease.berth?.pier.properties?.water ?? false,
    order,
  };
};
