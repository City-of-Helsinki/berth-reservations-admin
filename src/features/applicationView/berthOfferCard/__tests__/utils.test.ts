import { getOfferDetailsData } from '../utils';
import { lease } from '../__fixtures__/mockData';

describe('utils', () => {
  describe('getOfferDetailsData', () => {
    test('should return the correct data', () => {
      expect(getOfferDetailsData(lease)).toMatchSnapshot();
    });
  });
});
