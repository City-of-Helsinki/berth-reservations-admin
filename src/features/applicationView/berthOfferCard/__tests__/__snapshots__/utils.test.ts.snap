// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils getOfferDetailsData should return the correct data 1`] = `
Object {
  "berthComment": "",
  "berthDepth": null,
  "berthIsAccessible": false,
  "berthLength": 6,
  "berthMooringType": "SINGLE_SLIP_PLACE",
  "berthNum": "1",
  "berthWidth": 2.5,
  "customerEmail": "test@example.com",
  "electricity": true,
  "gate": true,
  "harborName": "Aurinkolahden venesatama (Aurinkoranta)",
  "id": "165023f5-5b74-4656-9315-d4e1a22f111d",
  "lighting": true,
  "order": Object {
    "fixedProducts": Array [
      Object {
        "id": "dc7f8158-a4cb-49a9-9368-e586f224fdb5",
        "name": "ELECTRICITY",
        "orderId": "9b97b9bc-bba5-46d9-aae2-86e8686012ef",
        "price": "20.50",
      },
      Object {
        "id": "2e55f605-061e-4299-814d-bf2cd53aa8ce",
        "name": "WATER",
        "orderId": "ea2ef175-f790-45a0-b16f-9e5a98340691",
        "price": "19.33",
      },
      Object {
        "id": "3a2e3a2f-2782-40e9-a139-b9a4d4a36c3b",
        "name": "GATE",
        "orderId": "8ba4de61-63cc-49fc-afe2-26068f3c8e4d",
        "price": "2.42",
      },
      Object {
        "id": "762da6e9-89be-43bc-b097-53dff4ed8415",
        "name": "MOORING",
        "orderId": "59c768c1-f371-4178-afda-1e16d2c725b0",
        "price": "8.33",
      },
      Object {
        "id": "34c5b6bc-7eb8-457e-bce8-8ba3a07b9057",
        "name": "WASTE_COLLECTION",
        "orderId": "6d910d8e-80b6-4486-aa85-4958bcd98dd5",
        "price": "4.17",
      },
      Object {
        "id": "650dbfcb-8f10-46b6-af64-197b77686a5b",
        "name": "LIGHTING",
        "orderId": "ab7cf1e5-92b5-42ad-844a-e2d577d30baf",
        "price": "40.28",
      },
    ],
    "id": "07f65f54-4a25-47fd-9eea-d9fdddfa46f9",
    "optionalProducts": Array [
      Object {
        "id": "5194dd14-0226-4603-808c-831cfbec72de",
        "name": "DINGHY_PLACE",
        "orderId": "244a7af9-6f91-4f08-9738-82ef06a2e790",
        "price": "65.00",
      },
    ],
    "orderNumber": "123456",
    "price": "96.67",
    "status": "WAITING",
    "totalPrice": "256.70",
  },
  "pierIdentifier": "C",
  "status": "DRAFTED",
  "wasteCollection": true,
  "water": true,
}
`;
